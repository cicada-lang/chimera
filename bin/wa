#!/usr/bin/env -S deno run --allow-all

import { createCommandRunner } from "../src/console/index.ts"

createCommandRunner().run()
