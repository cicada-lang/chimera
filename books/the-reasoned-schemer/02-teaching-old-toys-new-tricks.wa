Car [p, a]
---------- {
  [a, ...d] = p
}

query q {
  Car [["a", "c", "o", "r", "n"], q]
}

query q {
  Car [["a", "c", "o", "r", "n"], "a"]
}

query r {
  Car [[r, ...y], x]
  "pear" = x
}

query r {
  Car [["grape", "raisin", "pear"], x]
  Car [[["a"], ["b"], ["c"]], y]
  [x, ...y] = r
}

Cdr [p, d]
---------- {
  [a, ...d] = p
}

query r {
  Cdr [["a", "c", "o", "r", "n"], v]
  Cdr [v, w]
  Car [w, r]
}

query r {
  Cdr [["grape", "raisin", "pear"], x]
  Car [[["a"], ["b"], ["c"]], y]
  [x, ...y] = r
}

query q {
  Cdr [["a", "c", "o", "r", "n"], ["c", "o", "r", "n"]]
}

query x {
  Cdr [["c", "o", "r", "n"], [x, "r", "n"]]
}

query l {
  Cdr [l, ["c", "o", "r", "n"]]
  Car [l, x]
  x = "a"
}

// Cons [a, d, p]
// --------------- {
//   [a, ...d] = p
// }

Cons [a, d, p]
--------------- {
  Car [p, a]
  Cdr [p, d]
}

query l {
  Cons [["a", "b", "c"], ["d", "e"], l]
}

query x {
  Cons [x, ["a", "b", "c"], ["d", "a", "b", "c"]]
}

query r {
  ["e", "a", "d", x] = r
  Cons [y, ["a", z, "c"], r]
}

query x {
  Cons [x, ["a", x, "c"], ["d", "a", x, "c"]]
}

query l {
  ["d", "a", x, "c"] = l
  Cons [x, ["a", x, "c"], l]
}

query l {
  Cons [x, ["a", x, "c"], l]
  ["d", "a", x, "c"] = l
}

query l {
  Cons [w, ["n", "u", "s"], t]
  Cdr [l, t]
  Car [l, x]
  "b" = x
  Cdr [l, d]
  Car [d, y]
  "o" = y
}

Null x
---------- {
  [] = x
}

query q {
  Null ["grape", "raisin", "pear"]
}

query q {
  Null []
}

query x {
  Null x
}

query r {
  [x, y, "salad"] = r
}

Pair p
------ {
  Cons [a, d, p]
}

query q {
  Pair [q, ...q]
}

query q {
  Pair []
}

query q {
  Pair "pair"
}

query x {
  Pair x
}

query r {
  Pair [r]
}

// Singleton l
// ----------- {
//   Pair l
//   Cdr [l, d]
//   Null d
// }

Singleton l
----------- {
  Cdr [l, d]
  Null d
}

query r {
  Singleton ["tofu"]
}

query r {
  Singleton [["tofu"]]
}

query r {
  Singleton "tofu"
}

query r {
  Singleton ["e", "tofu"]
}

query r {
  Singleton []
}

query r {
  Singleton ["e", ..."tofu"]
}
