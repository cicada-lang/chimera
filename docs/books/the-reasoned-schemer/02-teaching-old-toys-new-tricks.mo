export clause Car(p, a)
---------- {
  Equal([a | _d], p)
}

print find q {
  Car(["a", "c", "o", "r", "n"], q)
}

print find _ {
  Car(["a", "c", "o", "r", "n"], "a")
}

print find r {
  Car([r | _y], x)
  Equal("pear", x)
}

print find r {
  Car(["grape", "raisin", "pear"], x)
  Car([["a"], ["b"], ["c"]], y)
  Equal([x | y], r)
}

export clause Cdr(p, d)
---------- {
  Equal([_a | d], p)
}

print find r {
  Cdr(["a", "c", "o", "r", "n"], v)
  Cdr(v, w)
  Car(w, r)
}

print find r {
  Cdr(["grape", "raisin", "pear"], x)
  Car([["a"], ["b"], ["c"]], y)
  Equal([x | y], r)
}

print find _ {
  Cdr(["a", "c", "o", "r", "n"], ["c", "o", "r", "n"])
}

print find x {
  Cdr(["c", "o", "r", "n"], [x, "r", "n"])
}

print find l {
  Cdr(l, ["c", "o", "r", "n"])
  Car(l, x)
  Equal(x, "a")
}

// clause Cons(a, d, p)
// --------------- {
//   Equal([a | d], p)
// }

export clause Cons(a, d, p)
--------------- {
  Car(p, a)
  Cdr(p, d)
}

print find l {
  Cons(["a", "b", "c"], ["d", "e"], l)
}

print find x {
  Cons(x, ["a", "b", "c"], ["d", "a", "b", "c"])
}

print find r {
  Equal(["e", "a", "d", _x], r)
  Cons(_y, ["a", _z, "c"], r)
}

print find x {
  Cons(x, ["a", x, "c"], ["d", "a", x, "c"])
}

print find l {
  Equal(["d", "a", x, "c"], l)
  Cons(x, ["a", x, "c"], l)
}

print find l {
  Cons(x, ["a", x, "c"], l)
  Equal(["d", "a", x, "c"], l)
}

print find l {
  Cons(_w, ["n", "u", "s"], t)
  Cdr(l, t)
  Car(l, x)
  Equal("b", x)
  Cdr(l, d)
  Car(d, y)
  Equal("o", y)
}

export clause Null(x)
---------- {
  Equal([], x)
}

print find _ {
  Null(["grape", "raisin", "pear"])
}

print find _ {
  Null([])
}

print find x {
  Null(x)
}

print find r {
  Equal([_x, _y, "salad"], r)
}

clause Pair(p)
------ {
  Cons(_a, _d, p)
}

print find q {
  Pair([q | q])
}

print find _ {
  Pair([])
}

print find _ {
  Pair("pair")
}

print find x {
  Pair(x)
}

print find r {
  Pair([r])
}

// clause Singleton(l)
// ----------- {
//   Pair(l)
//   Cdr(l, d)
//   Null(d)
// }

clause Singleton(l)
----------- {
  Cdr(l, d)
  Null(d)
}

print find _ {
  Singleton(["tofu"])
}

print find _ {
  Singleton([["tofu"]])
}

print find _ {
  Singleton("tofu")
}

print find _ {
  Singleton(["e", "tofu"])
}

print find _ {
  Singleton([])
}

print find _ {
  Singleton(["e" | "tofu"])
}
