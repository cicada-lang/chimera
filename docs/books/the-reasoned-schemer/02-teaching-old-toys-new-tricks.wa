Car [p, a]
---------- {
  [a | _d] = p
}

find q {
  Car [["a", "c", "o", "r", "n"], q]
}

find q {
  Car [["a", "c", "o", "r", "n"], "a"]
}

find r {
  Car [[r | y], x]
  "pear" = x
}

find r {
  Car [["grape", "raisin", "pear"], x]
  Car [[["a"], ["b"], ["c"]], y]
  [x | y] = r
}

Cdr [p, d]
---------- {
  [_a | d] = p
}

find r {
  Cdr [["a", "c", "o", "r", "n"], v]
  Cdr [v, w]
  Car [w, r]
}

find r {
  Cdr [["grape", "raisin", "pear"], x]
  Car [[["a"], ["b"], ["c"]], y]
  [x | y] = r
}

find q {
  Cdr [["a", "c", "o", "r", "n"], ["c", "o", "r", "n"]]
}

find x {
  Cdr [["c", "o", "r", "n"], [x, "r", "n"]]
}

find l {
  Cdr [l, ["c", "o", "r", "n"]]
  Car [l, x]
  x = "a"
}

// Cons [a, d, p]
// --------------- {
//   [a | d] = p
// }

Cons [a, d, p]
--------------- {
  Car [p, a]
  Cdr [p, d]
}

find l {
  Cons [["a", "b", "c"], ["d", "e"], l]
}

find x {
  Cons [x, ["a", "b", "c"], ["d", "a", "b", "c"]]
}

find r {
  ["e", "a", "d", x] = r
  Cons [y, ["a", z, "c"], r]
}

find x {
  Cons [x, ["a", x, "c"], ["d", "a", x, "c"]]
}

find l {
  ["d", "a", x, "c"] = l
  Cons [x, ["a", x, "c"], l]
}

find l {
  Cons [x, ["a", x, "c"], l]
  ["d", "a", x, "c"] = l
}

find l {
  Cons [w, ["n", "u", "s"], t]
  Cdr [l, t]
  Car [l, x]
  "b" = x
  Cdr [l, d]
  Car [d, y]
  "o" = y
}

Null x
---------- {
  [] = x
}

find q {
  Null ["grape", "raisin", "pear"]
}

find q {
  Null []
}

find x {
  Null x
}

find r {
  [x, y, "salad"] = r
}

Pair p
------ {
  Cons [_a, _d, p]
}

find q {
  Pair [q | q]
}

find q {
  Pair []
}

find q {
  Pair "pair"
}

find x {
  Pair x
}

find r {
  Pair [r]
}

// Singleton l
// ----------- {
//   Pair l
//   Cdr [l, d]
//   Null d
// }

Singleton l
----------- {
  Cdr [l, d]
  Null d
}

find r {
  Singleton ["tofu"]
}

find r {
  Singleton [["tofu"]]
}

find r {
  Singleton "tofu"
}

find r {
  Singleton ["e", "tofu"]
}

find r {
  Singleton []
}

find r {
  Singleton ["e" | "tofu"]
}
