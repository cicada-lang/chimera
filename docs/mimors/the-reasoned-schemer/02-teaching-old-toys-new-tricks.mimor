<theme color="blue" />

<question>
  How do we translate a function into a relation?

  <answer>
    The Translation (Initial)

    To translate a function into a relation,
    first replace define with defrel.

    Then unnest each expression in each cond line,
    and replace each cond with conde.

    To unnest a #t, replace it with #s.

    To unnest a #f, replace it with #u.
  </answer>
</question>

<question>
  The Law of #u (fail)

  <answer>
    Any conde line that has #u as a top-level goal cannot contribute values.
  </answer>
</question>
