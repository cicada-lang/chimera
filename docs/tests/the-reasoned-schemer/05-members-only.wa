import { Null, Cons, Car, Cdr } from "./02-teaching-old-toys-new-tricks.wa"

Mem [x, l, out]
--------------- {
  Car [l, x]
  unify out = l
}

Mem [x, l, out]
--------------- {
  Cdr [l, d]
  Mem [x, d, out]
}

query x {
  Mem [
    "fig",
    ["roll", "okra", "fig", "beet", "roll", "pea"],
    x,
  ]
}

query x {
  Mem [
    "fig",
    ["roll", "okra", "beet", "beet", "roll", "pea"],
    x,
  ]
}

query x {
  Mem ["fig", ["pea"], ["pea"]]
}

query out {
  Mem ["fig", ["fig"], out]
}

query out {
  Mem ["fig", ["fig", "pea"], out]
}

query x {
  Mem [
    x,
    ["roll", "okra", "fig", "beet", "roll", "pea"],
    ["fig", "beet", "roll", "pea"],
  ]
}

query x {
  Mem ["fig", ["fig", "pea"], ["pea", x]]
}

query x {
  Mem ["fig", ["fig", "pea"], [x, "pea"]]
}
