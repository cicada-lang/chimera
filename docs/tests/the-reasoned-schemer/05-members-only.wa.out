[ "a" ]
[ "_.0" ]
[ "pear" ]
[ ["grape", "a"] ]
[ "o" ]
[ [["raisin", "pear"], "a"] ]
[ "_.0" ]
[ "o" ]
[ ["a", "c", "o", "r", "n"] ]
[ [["a", "b", "c"], "d", "e"] ]
[ "d" ]
[ ["e", "a", "d", "c"] ]
[ "d" ]
[ ["d", "a", "d", "c"] ]
[ ["d", "a", "d", "c"] ]
[ ["b", "o", "n", "u", "s"] ]
[]
[ "_.0" ]
[ [] ]
[ ["_.0", "_.1", "salad"] ]
[ "_.0" ]
[]
[]
[ ["_.0", { "...": "_.1" }] ]
[ "_.0" ]
[ "_.0" ]
[ "_.0" ]
[]
[]
[]
[]
[ ["fig", "beet", "roll", "pea"] ]
[]
[]
[ ["fig"] ]
[ ["fig", "pea"] ]
[ "fig" ]
[]
[ "fig" ]
